### PR TITLE
Add state to mapcaps

### PR DESCRIPTION
**Pull request by gfiorav**
Fixes https://github.com/CartoDB/cartodb/issues/8463
Fixes https://github.com/CartoDB/cartodb/issues/8464

### CODE REVIEW

**Comment by gfiorav** [comment]
Ok! CI is 🍏 

This persists the builder state, allowing you to share a map with a given widget filter applied and that same configuration will load in the embed by default. If you overwrite the state, the url will update as usual and you will be able to share that new state by copying the link.

Basically, in addition to `vizJSON`, there's now `stateJSON`, that is used for this purpose.

@xavijam please CR the front part.
@juanignaciosl please CR the back part.

cc/ @saleiva @javisantana @fdansv

File: lib/assets/javascripts/cartodb3/helpers/url-state-helper.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ *   URLStateHelper is ported from cartodb-deep-insights.js
+ *                                 github.com/CartoDB/deep-insights.js
+ *
+ *   Original author is Francisco Dans
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/helpers/url-state-helper.js]: I feel we don't need this :joy:

File: app/controllers/carto/api/mapcap_presenter.rb [hunk 2]
```diff
@@ -13,7 +13,8 @@ def to_poro
 
         {
           id: @mapcap.id,
-          related_ids: @mapcap.ids_json,
+          ids_json: @mapcap.ids_json,
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcap_presenter.rb]: How come `related_ids` was not previously used?

File: db/migrate/20160704184357_add_state_to_mapcaps.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Sequel.migration do
+  up do
+    alter_table :mapcaps do
+      add_column :state_json, :string, type: 'json', null: false, default: '{}'
```
Comment (juanignaciosl) [inline db/migrate/20160704184357_add_state_to_mapcaps.rb]: Please don't add the type to the name. I know this is a big change, but... U_U

File: lib/assets/javascripts/cartodb3/helpers/url-state-helper.js (hunk 1 above)
Comment (fdansv) [inline lib/assets/javascripts/cartodb3/helpers/url-state-helper.js]: YES WE TOTALLY NEED IT

**Comment by juanignaciosl** [comment]
I've made a major comment about column name. I know that the change is a PITA but as database has json type and you use there's no reason to add it to the name. 👍 afterwards.

File: lib/assets/javascripts/cartodb3/helpers/url-state-helper.js (hunk 1 above)
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/helpers/url-state-helper.js]: I didn't know you could invoke a file from another repo, totally doing that!

**Comment by gfiorav** [comment]
Not happening! I'll come back with a state-everywhere solution. Thanks for the comments @juanignaciosl ❤️

**Pull request closed without merging**